File: screen-switcher/src/test/java/com/jaynewstrom/screenswitcher/ScreenSwitcherStateTest.java
package com.jaynewstrom.screenswitcher;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public final class ScreenSwitcherStateTest {

    @Test public void constructorMakesDefensiveCopyOfScreensPassedIn() {
        List<Screen> passedList = Collections.singletonList(mock(Screen.class));
        ScreenSwitcherState state = ScreenTestUtils.defaultState(passedList);
        state.getScreens().add(mock(Screen.class));
        assertThat(passedList).hasSize(1);
        assertThat(state.getScreens()).hasSize(2);
    }

    @Test public void constructorDoesNotAllowNulScreens() {
        try {
            //noinspection ConstantConditions
            ScreenTestUtils.defaultState((List<Screen>) null);
            fail();
        } catch (NullPointerException expected) {
            assertThat(expected).hasMessage("screens == null");
        }
    }

    @Test public void constructorDoesNotAllowEmptyScreens() {
        try {
            ScreenTestUtils.defaultState(Collections.<Screen>emptyList());
            fail();
        } catch (IllegalArgumentException expected) {
            assertThat(expected).hasMessage("screens must contain at least one screen");
        }
    }

    @Test public void constructorRejectsNullInTheListOfScreens() {
        try {
            ScreenTestUtils.defaultState(Collections.<Screen>singletonList(null));
            fail();
        } catch (NullPointerException expected) {
            assertThat(expected).hasMessage("screen == null");
        }
    }

    @Test public void constructorRejectsDuplicateScreens() {
        Screen screen = mock(Screen.class);
        try {
            ScreenTestUtils.defaultState(Arrays.asList(screen, screen));
            fail();
        } catch (IllegalArgumentException expected) {
            assertThat(expected).hasMessage("screen already exists");
        }
    }

    @Test public void handlesPopIsFalseWhenNoPopListeners() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        assertThat(state.handlesPop(screen)).isFalse();
    }

    @Test public void handlesPopIsFalseWhenPopListenerReturnsFalse() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        ScreenPopListener popListener = mock(ScreenPopListener.class);
        state.registerPopListener(screen, popListener);
        when(popListener.onScreenPop(screen)).thenReturn(false);
        assertThat(state.handlesPop(screen)).isFalse();
    }

    @Test public void handlesPopIsTrueWhenPopListenerReturnsTrue() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        ScreenPopListener popListener = mock(ScreenPopListener.class);
        state.registerPopListener(screen, popListener);
        when(popListener.onScreenPop(screen)).thenReturn(true);
        assertThat(state.handlesPop(screen)).isTrue();
    }

    @Test public void handlesPopIsFalseForScreenNotMatchingPopListenerThatReturnsTrue() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        ScreenPopListener popListener = mock(ScreenPopListener.class);
        state.registerPopListener(screen, popListener);
        when(popListener.onScreenPop(screen)).thenReturn(true);
        assertThat(state.handlesPop(mock(Screen.class))).isFalse();
    }

    @Test public void handlesPopRemovesPopListenerWhenReturnsFalse() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        ScreenPopListener popListener = mock(ScreenPopListener.class);
        state.registerPopListener(screen, popListener);
        when(popListener.onScreenPop(screen)).thenReturn(false);
        assertThat(state.handlesPop(screen)).isFalse(); // Will be removed here.
        assertThat(state.handlesPop(screen)).isFalse(); // Will return false because it's gone.
        verify(popListener).onScreenPop(screen); // Will verify it was only used once, even though #handlesPop was called twice.
    }

    @Test public void handlesPopKeepsPopListenerWhenReturnsTrue() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        ScreenPopListener popListener = mock(ScreenPopListener.class);
        state.registerPopListener(screen, popListener);
        when(popListener.onScreenPop(screen)).thenReturn(true);
        assertThat(state.handlesPop(screen)).isTrue();
        assertThat(state.handlesPop(screen)).isTrue();
        verify(popListener, times(2)).onScreenPop(screen);
    }

    @Test public void addScreenRejectsDuplicateScreen() {
        Screen screen = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(screen);
        try {
            state.addScreen(screen);
            fail();
        } catch (IllegalArgumentException expected) {
            assertThat(expected).hasMessage("screen already exists");
        }
    }

    @Test public void addScreenRejectsNullScreen() {
        ScreenSwitcherState state = ScreenTestUtils.defaultState(mock(Screen.class));
        try {
            state.addScreen(null);
            fail();
        } catch (NullPointerException expected) {
            assertThat(expected).hasMessage("screen == null");
        }
    }

    @Test public void indexOfRejectsNullScreen() {
        ScreenSwitcherState state = ScreenTestUtils.defaultState(mock(Screen.class));
        try {
            //noinspection ConstantConditions
            state.indexOf(null);
            fail();
        } catch (NullPointerException expected) {
            assertThat(expected).hasMessage("screen == null");
        }
    }

    @Test public void indexOfReturnsNegativeOneIfTheScreenDoesNotExist() {
        ScreenSwitcherState state = ScreenTestUtils.defaultState(mock(Screen.class));
        assertThat(state.indexOf(mock(Screen.class))).isEqualTo(-1);
    }

    @Test public void indexOfReturnsTheCorrectIndexOfAnExistingScreen() {
        Screen screen0 = mock(Screen.class);
        Screen screen1 = mock(Screen.class);
        Screen screen2 = mock(Screen.class);
        ScreenSwitcherState state = ScreenTestUtils.defaultState(Arrays.asList(screen0, screen1, screen2));
        assertThat(state.indexOf(screen1)).isEqualTo(1);
    }

    @Test public void screenCountIsUpdatedWhenScreensAreAdded() {
        ScreenSwitcherState state = ScreenTestUtils.defaultState(Arrays.asList(mock(Screen.class), mock(Screen.class)));
        assertThat(state.screenCount()).isEqualTo(2);
        state.addScreen(mock(Screen.class));
        assertThat(state.screenCount()).isEqualTo(3);
    }
}

File: screen-switcher/src/test/java/com/jaynewstrom/screenswitcher/ScreenSwitcherStateTest.kt
package com.jaynewstrom.screenswitcher

import android.os.Parcelable
import android.util.SparseArray
import android.view.View
import org.fest.assertions.api.Assertions.assertThat
import org.fest.assertions.data.MapEntry
import org.junit.Assert.fail
import org.junit.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import java.util.concurrent.atomic.AtomicInteger

class ScreenSwitcherStateTest {
    @Test fun constructorMakesDefensiveCopyOfScreensPassedIn() {
        val passedList = listOf(mock(Screen::class.java))
        val state = ScreenTestUtils.defaultState(passedList)
        state.screens.add(mock(Screen::class.java))
        assertThat(passedList).hasSize(1)
        assertThat(state.screens).hasSize(2)
    }

    @Test fun constructorDoesNotAllowEmptyScreens() {
        try {
            ScreenTestUtils.defaultState(emptyList())
            fail()
        } catch (expected: IllegalArgumentException) {
            assertThat(expected).hasMessage("screens must contain at least one screen")
        }
    }

    @Test fun constructorRejectsDuplicateScreens() {
        val screen = mock(Screen::class.java)
        try {
            ScreenTestUtils.defaultState(listOf(screen, screen))
            fail()
        } catch (expected: IllegalArgumentException) {
            assertThat(expected).hasMessage("screen already exists")
        }
    }

    @Test fun handlesPopIsFalseWhenNoPopListeners() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        assertThat(state.handlesPop(view, screen, null)).isFalse
    }

    @Test fun setPopListenerSkipsMissingScreens() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListener = mock(ScreenPopListener::class.java)
        val missingScreen = mock(Screen::class.java)
        assertThat(state.setPopListener(missingScreen, popListener)).isFalse
        assertThat(state.handlesPop(view, screen, null)).isFalse
        verify(popListener, never()).onScreenPop(kotlinAny(), kotlinAny())
    }

    @Test fun handlesPopIsFalseWhenPopListenerReturnsFalse() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListener = mock(ScreenPopListener::class.java)
        state.setPopListener(screen, popListener)
        `when`(popListener.onScreenPop(view, null)).thenReturn(false)
        assertThat(state.handlesPop(view, screen, null)).isFalse
    }

    @Test fun handlesPopIsTrueWhenPopListenerReturnsTrue() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListener = mock(ScreenPopListener::class.java)
        state.setPopListener(screen, popListener)
        `when`(popListener.onScreenPop(view, null)).thenReturn(true)
        assertThat(state.handlesPop(view, screen, null)).isTrue
    }

    @Test fun handlesPopIsFalseForScreenNotMatchingPopListenerThatReturnsTrue() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListener = mock(ScreenPopListener::class.java)
        state.setPopListener(screen, popListener)
        `when`(popListener.onScreenPop(view, null)).thenReturn(true)
        assertThat(state.handlesPop(mock(View::class.java), mock(Screen::class.java), null)).isFalse
    }

    @Test fun removeScreenRemovesPopListener() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListener = mock(ScreenPopListener::class.java)
        state.setPopListener(screen, popListener)
        `when`(popListener.onScreenPop(view, null)).thenReturn(true)
        assertThat(state.handlesPop(view, screen, null)).isTrue
        state.removeScreen(screen)
        assertThat(state.handlesPop(view, screen, null)).isFalse
    }

    @Test fun handlesPopKeepsPopListenerWhenReturnsTrue() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListener = mock(ScreenPopListener::class.java)
        state.setPopListener(screen, popListener)
        `when`(popListener.onScreenPop(view, null)).thenReturn(true)
        assertThat(state.handlesPop(view, screen, null)).isTrue
        assertThat(state.handlesPop(view, screen, null)).isTrue
        verify(popListener, times(2)).onScreenPop(view, null)
    }

    @Test fun handlesPopPassesPopContextToPopListener() {
        val screen = mock(Screen::class.java)
        val view = mock(View::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val popListenerCalledCount = AtomicInteger(0)
        val popListener = object : ScreenPopListener {
            override fun onScreenPop(view: View, popContext: Any?): Boolean {
                @Suppress("UNCHECKED_CAST")
                assertThat(popContext as Map<String, String>).contains(MapEntry.entry("Foo", "Bar"))
                popListenerCalledCount.incrementAndGet()
                return false
            }
        }
        state.setPopListener(screen, popListener)
        val popContext = mapOf(Pair("Foo", "Bar"))
        assertThat(state.handlesPop(view, screen, popContext)).isFalse
        assertThat(popListenerCalledCount.get()).isEqualTo(1)
    }

    @Test fun addScreenRejectsDuplicateScreen() {
        val screen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        try {
            state.addScreen(screen)
            fail()
        } catch (expected: IllegalArgumentException) {
            assertThat(expected).hasMessage("screen already exists")
        }
    }

    @Test fun indexOfReturnsNegativeOneIfTheScreenDoesNotExist() {
        val state = ScreenTestUtils.defaultState(mock(Screen::class.java))
        assertThat(state.indexOf(mock(Screen::class.java))).isEqualTo(-1)
    }

    @Test fun indexOfReturnsTheCorrectIndexOfAnExistingScreen() {
        val screen0 = mock(Screen::class.java)
        val screen1 = mock(Screen::class.java)
        val screen2 = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(listOf(screen0, screen1, screen2))
        assertThat(state.indexOf(screen1)).isEqualTo(1)
    }

    @Test fun screenCountIsUpdatedWhenScreensAreAdded() {
        val state = ScreenTestUtils.defaultState(listOf(mock(Screen::class.java), mock(Screen::class.java)))
        assertThat(state.screenCount()).isEqualTo(2)
        state.addScreen(mock(Screen::class.java))
        assertThat(state.screenCount()).isEqualTo(3)
    }

    @Test fun removeActiveScreenTransitionReturnsTheCorrectTransitionForScreen() {
        val screen0 = mock(Screen::class.java)
        val transition0: (data: ScreenSwitcherData) -> Unit = { throw AssertionError("0") }
        val screen1 = mock(Screen::class.java)
        val transition1: (data: ScreenSwitcherData) -> Unit = { throw AssertionError("1") }
        val screen2 = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(listOf(screen0, screen1, screen2))
        state.enqueueTransition(screen0, transition0)
        state.enqueueTransition(screen1, transition1)
        assertThat(state.removeActiveScreenTransition()).isNull() // 2
        state.removeScreen(screen2)
        assertThat(state.removeActiveScreenTransition()).isSameAs(transition1) // 1
        assertThat(state.removeActiveScreenTransition()).isNull() // 1 is gone now
        state.removeScreen(screen1)
        assertThat(state.removeActiveScreenTransition()).isSameAs(transition0) // 0
        state.removeScreen(screen0)
        assertThat(state.removeActiveScreenTransition()).isNull() // None left
    }

    @Test fun viewHierarchyStateCanBeAddedAndRemoved() {
        val screen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val savedHierarchy = SparseArray<Parcelable>()
        state.saveViewHierarchyState(screen, savedHierarchy)
        assertThat(state.removeViewHierarchyState(screen)).isEqualTo(savedHierarchy)
        assertThat(state.removeViewHierarchyState(screen)).isNull() // It's gone!
    }

    @Test fun removeScreenRemovesViewHierarchyState() {
        val screen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(screen)
        val savedHierarchy = SparseArray<Parcelable>()
        state.saveViewHierarchyState(screen, savedHierarchy)
        state.removeScreen(screen)
        assertThat(state.removeViewHierarchyState(screen)).isNull() // It's gone!
    }

    @Test fun createNestedStateCopiesLifecycleListener() {
        val rootScreen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(rootScreen)
        val nestedScreen = mock(Screen::class.java)
        val nestedState = state.createNestedState(listOf(nestedScreen))
        assertThat(state.lifecycleListener).isSameAs(nestedState.lifecycleListener)
    }

    @Test fun testScreenSwitcherCreatedListenerNotAddedWhenScreenDoesNotExist() {
        val rootScreen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(rootScreen)
        val associatedScreen = mock(Screen::class.java)
        val listener = mock(ScreenSwitcherState.ScreenSwitcherCreatedListener::class.java)
        assertThat(state.registerScreenSwitcherCreatedListener(associatedScreen, listener)).isFalse
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
        verify(listener, never()).screenSwitcherCreated(kotlinAny())
        assertThat(state.unregisterScreenSwitcherCreatedListener(associatedScreen, listener)).isFalse
    }

    @Test fun testScreenSwitcherCreatedListenerHappyPath() {
        val rootScreen = mock(Screen::class.java)
        val associatedScreen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(listOf(rootScreen, associatedScreen))
        val listener = mock(ScreenSwitcherState.ScreenSwitcherCreatedListener::class.java)
        assertThat(state.registerScreenSwitcherCreatedListener(associatedScreen, listener)).isTrue
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
        verify(listener).screenSwitcherCreated(kotlinAny())
        verifyNoMoreInteractions(listener)
        assertThat(state.unregisterScreenSwitcherCreatedListener(associatedScreen, listener)).isTrue
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
        assertThat(state.unregisterScreenSwitcherCreatedListener(associatedScreen, listener)).isFalse
    }

    @Test fun testScreenSwitcherCreatedListenerMultipleListenersCalled() {
        val rootScreen = mock(Screen::class.java)
        val associatedScreen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(listOf(rootScreen, associatedScreen))
        val listener1 = mock(ScreenSwitcherState.ScreenSwitcherCreatedListener::class.java)
        val listener2 = mock(ScreenSwitcherState.ScreenSwitcherCreatedListener::class.java)
        assertThat(state.registerScreenSwitcherCreatedListener(associatedScreen, listener1)).isTrue
        assertThat(state.registerScreenSwitcherCreatedListener(associatedScreen, listener2)).isTrue
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
        verify(listener1).screenSwitcherCreated(kotlinAny())
        verify(listener2).screenSwitcherCreated(kotlinAny())
        assertThat(state.unregisterScreenSwitcherCreatedListener(associatedScreen, listener1)).isTrue
        assertThat(state.unregisterScreenSwitcherCreatedListener(associatedScreen, listener2)).isTrue
        verifyNoMoreInteractions(listener1, listener2)
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
    }

    @Test fun testScreenSwitcherCreatedListenerIsRemovedWhenScreenIsRemoved() {
        val rootScreen = mock(Screen::class.java)
        val associatedScreen = mock(Screen::class.java)
        val state = ScreenTestUtils.defaultState(listOf(rootScreen, associatedScreen))
        val listener = mock(ScreenSwitcherState.ScreenSwitcherCreatedListener::class.java)
        assertThat(state.registerScreenSwitcherCreatedListener(associatedScreen, listener)).isTrue
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
        verify(listener).screenSwitcherCreated(kotlinAny())
        verifyNoMoreInteractions(listener)
        state.removeScreen(associatedScreen)
        state.screenSwitcherCreated(mock(ScreenSwitcher::class.java))
        assertThat(state.unregisterScreenSwitcherCreatedListener(associatedScreen, listener)).isFalse
    }
}
